### PR TITLE
ref: new workspace structure and version management

### PR DESCRIPTION
**Pull request by bidzyyys**
- [x]  Set `motsu` version to `v0.1.0`
- [x] Improve version management for the workspace
  - `contracts`, `lib/crypto`, `examples`, and `benches` use version set for the whole workspace 
  - `motsu`, `motsu-proc`, `e2e`, `e2e-proc` have their own version per crate
- [x] Cleanup dependencies for the workspace
- [x] Add CI checks for pre-publishing on crates.io
- [x] Fix issues with publishing `motsu` to crates.io:
```
cargo publish -p motsu --target wasm32-unknown-unknown --dry-run
    Updating crates.io index
error: all dependencies must have a version specified when publishing.
dependency `motsu-proc` does not specify a version
Note: The published dependency will use the version from crates.io,
the `path` specification will be removed from the dependency declaration.
```

### CODE REVIEW

File: lib/motsu/Cargo.toml [hunk 1]
```diff
@@ -5,14 +5,14 @@ edition.workspace = true
...
-version = "0.1.0"
+version = "0.1.0-rc"
 
 [dependencies]
 const-hex.workspace = true
 once_cell.workspace = true
 tiny-keccak.workspace = true
 stylus-sdk.workspace = true
-motsu-proc = { path = "../motsu-proc" }
+motsu-proc = "0.1.0"
```
Comment (qalisander) [inline lib/motsu/Cargo.toml:15]: I think If we change smth in `motsu-proc` crate we need to see those changes inside `motsu` without publishing crate

Comment (qalisander) [reply lib/motsu/Cargo.toml:15]: Can you try to add version altogether with path?

File: .github/workflows/check-publish.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    branches: [ main ]
+  pull_request:
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+env:
+  CARGO_TERM_COLOR: always
+jobs:
+  check-publish:
+    name: Check publish on crates.io
```
Comment (qalisander) [inline .github/workflows/check-publish.yml:16]: That looks cool!

**Review by qalisander — APPROVED** [review]
Good job!
Definitely looks better

File: .github/workflows/check-publish.yml (hunk 2 above)
Comment (qalisander) [reply .github/workflows/check-publish.yml:16]: But I would change the pr's name to `ref: ...`

Comment (bidzyyys) [reply .github/workflows/check-publish.yml:16]: @qalisander how about `ref: new workspace structure`

**Review by ggonzalez94 — COMMENTED** [review]
Looking great! Just left a small comment

File: Cargo.toml [hunk 3]
```diff
@@ -95,6 +96,14 @@ syn = { version = "2.0.58", features = ["full"] }
 proc-macro2 = "1.0.79"
 quote = "1.0.35"
 
+# members
+openzeppelin-stylus = { path = "contracts" }
+openzeppelin-crypto = { path = "lib/crypto" }
+motsu = { path = "lib/motsu"}
+motsu-proc = { path = "lib/motsu-proc", version = "0.1.0" }
```
Comment (ggonzalez94) [inline Cargo.toml:103]: Why do we set a specific version for `motsu-proc`? Is this intentional?

**Review by ggonzalez94 — COMMENTED** [review]
One final question

File: .github/workflows/check-publish.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: set up rust
+        uses: dtolnay/rust-toolchain@master
+        id: toolchain
+        with:
+          target: wasm32-unknown-unknown
+          components: rust-src
+          toolchain: nightly-2024-01-01
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply .github/workflows/check-publish.yml:27]: https://github.com/OpenZeppelin/rust-contracts-stylus/issues/294

File: Cargo.toml (hunk 3 above)
Comment (bidzyyys) [reply Cargo.toml:103]: Needed to publish `motsu` to crates.io; requirement of `cargo publish ...`

**Review by ggonzalez94 — APPROVED** [review]
LGTM

**Pull request merged**